src/shogun/preprocessor/DensePreprocessor.cpp

```diff
@@ -116,6 +116,27 @@ DensePreprocessor<ST>::inverse_transform(std::shared_ptr<Features> features, boo
 	return std::make_shared<DenseFeatures<ST>>(feat_matrix);
 }
 
+template <class ST>
+void DensePreprocessor<ST>::fit(std::shared_ptr<Features> f)
+{
+	require(f, "Features are not provided");
+
+	auto dense_features = f->as<DenseFeatures<ST>>();
+	auto num_features = dense_features->get_num_features();
+	require(num_features > 0, "Dimension of provided features {} must be positive", num_features);
+	
+	auto feature_matrix = dense_features->get_feature_matrix();
+	fit_impl(feature_matrix);
+	
+	m_fitted = true;
+}
+
+template <class ST>
+void DensePreprocessor<ST>::fit_impl(SGMatrix<ST> feature_matrix)
```

Comment:
```suggestion
void DensePreprocessor<ST>::fit_impl(const SGMatrix<ST>& feature_matrix)
```

Reply:
Shouldn't this be pure virtual and then make this is a noop, rather  `not_implemented` when it is implemented? The issue with throwing is that you might add this to a `Pipeline`, and this will throw an error because the `Pipeline` object will always do fit + transform.

Reply:
Purely virtual isn't an option because few preproc don't employ fit_impl method. This would cause compile error, I guess we could make it empty function then?

Reply:
yes, I think so

Reply:
> Shouldn't this be pure virtual and then make this is a noop, rather `not_implemented` when it is implemented? The issue with throwing is that you might add this to a `Pipeline`, and this will throw an error because the `Pipeline` object will always do fit + transform.

@karlnapf  this is why it doesn't make sense(from the second sentence).

Reply:
@gf712 , what I did instead was make the fit_impl a noop in ```DensePreprocessor```. This way we needn't make it noop in each and every subclass which doesn't use fit_impl, and those who use fit_impl simply override this :)

Reply:
sure, I think that should be fine :)

Reply:
how many subclasses dont have a `fit` method? I guess that would be very few (`LogPlusOne`, which else?)
I'd actually prefer to be as explicit as possible with this. I.e. making the method purely virtual in the base class, and then the classes that don't need it override with as a no-op. That seems to make the structure much more clear.

Reply:
A total of five(Pnorm, NormOne, LogPlusOne, Homogenous, SumOne) don't have fit method. 

I see your point, I drew my inspiration from the ```fit``` implementation we have on ```Transformer.h``` it is defined as noop, so those who don't use it get a noop others override it. We can go with your method as well, but I'd like to know why we didn't use this same methodology on ```fit```?

Reply:
Interesting, thx for pointing that out. In that case let's stick with what is there then for consistency reasons (i.e. with what you suggested)

Asking why something is the way it is can be difficult in shogun ... many authors, many opinions ;)